FWCore/Framework/interface/CallbackExternalWork.h

```diff
@@ -0,0 +1,199 @@
+//
+/**\class edm::eventsetup::CallbackExternalWork
+
+ Description: Functional object used as the 'callback' for the CallbackProxy
+
+ Usage: Produces data objects for ESProducers in EventSetup system
+
+*/
+//
+// Author:      W. David Dagenhart
+// Created:     27 February 2023
+
+#include <exception>
+#include <memory>
+#include <optional>
+#include <utility>
+#include <vector>
+
+#include "oneapi/tbb/task_group.h"
+
+#include "FWCore/Concurrency/interface/SerialTaskQueueChain.h"
+#include "FWCore/Concurrency/interface/WaitingTask.h"
+#include "FWCore/Concurrency/interface/WaitingTaskHolder.h"
+#include "FWCore/Concurrency/interface/WaitingTaskList.h"
+#include "FWCore/Concurrency/interface/WaitingTaskWithArenaHolder.h"
+#include "FWCore/Framework/interface/CallbackBase.h"
+#include "FWCore/Framework/interface/EventSetupRecordImpl.h"
+#include "FWCore/Utilities/interface/ConvertException.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "FWCore/Utilities/interface/Signal.h"
+#include "FWCore/ServiceRegistry/interface/ActivityRegistry.h"
+#include "FWCore/ServiceRegistry/interface/ESModuleCallingContext.h"
+#include "FWCore/ServiceRegistry/interface/ESParentContext.h"
+#include "FWCore/ServiceRegistry/interface/ServiceRegistry.h"
+#include "FWCore/ServiceRegistry/interface/ServiceToken.h"
+
+namespace edm {
+
+  class EventSetupImpl;
+
+  namespace eventsetup {
+
+    template <typename T,             //producer's type
+              typename TAcquireFunc,  //acquire functor type
+              typename TProduceFunc,  //produce functor type
+              typename TReturn,       //return type of the producer's method
+              typename TRecord,       //the record passed in as an argument
+              typename TDecorator     //allows customization using pre/post calls
+              = CallbackSimpleDecorator<TRecord>>
+    class CallbackExternalWork
+        : public CallbackBase<CallbackExternalWork<T, TAcquireFunc, TProduceFunc, TReturn, TRecord, TDecorator>,
+                              T,
+                              TProduceFunc,
+                              TReturn,
+                              TRecord,
+                              TDecorator> {
+    public:
+      using Base = CallbackBase<CallbackExternalWork, T, TProduceFunc, TReturn, TRecord, TDecorator>;
+
+      CallbackExternalWork(T* iProd, TProduceFunc iProduceFunc, unsigned int iID, const TDecorator& iDec = TDecorator())
+          : Base(iProd, std::make_shared<TProduceFunc>(std::move(iProduceFunc)), iID, iDec),
+            acquireFunction_(std::shared_ptr<TAcquireFunc>()) {}
+
+      CallbackExternalWork(T* iProd,
+                           TAcquireFunc iAcquireFunc,
+                           TProduceFunc iProduceFunc,
+                           unsigned int iID,
+                           const TDecorator& iDec = TDecorator())
+          : Base(iProd, std::make_shared<TProduceFunc>(std::move(iProduceFunc)), iID, iDec),
```

Comment:
How about calling the constructor just below this one instead (i.e. it is then just a forwarding constructor).

Reply:
Done.